NLSY/nlsy_lib.R

```diff
@@ -243,6 +243,95 @@ nlsy_get_highest_grade_completed_df = function()
     return(data)
 }
 
+#' Returns marriage status the last month of a given year
+nlsy_get_marstat_df = function()
+{
+    data = nlsydf |>
+        select(
+            id =        "PUBID_1997",
+            starts_with('MAR_STATUS_')
+        ) |>
+        pivot_longer(
+            starts_with('MAR_STATUS_'),
+            names_to=c('year', 'month'),
+            names_pattern='MAR_STATUS_(\\d\\d\\d\\d)\\.(\\d\\d)_XRND',
+            values_to='mar_status'
+        ) |>
+        mutate(
+            year=as.integer(year),
+            month=as.integer(month)
+        ) |>
+        filter(month == 12) |>
+        group_by(id) |>
+        complete(year=1997:2021) |>
+        fill(mar_status, .direction = "downup") |>
+    #Counting missings:sum(is.na(data))/12 = 203 out of 8984 in nlsydf
+        mutate(mar_status = case_when(
+            is.na(mar_status) ~ 'Never Married, Not Cohabitating',
+            TRUE ~ mar_status)
+         ) |>
+        mutate(
+            single=as.integer(mar_status == 'Never Married, Not Cohabitating'),
+            cohabitating=as.integer(mar_status == 'Never Married, Cohabiting'),
+            married=as.integer(mar_status == 'Married'),
+            divorced=as.integer(mar_status == 'Divorced'),
+            widowed=as.integer(mar_status == 'Widowed')
+        )|>
+        select(-month)
+
+    return(data)
+}
+
+#' Returns number of biological children in and out of household
+nlsy_get_biochild_df = function()
+{
+    data1 = nlsydf |>
+        select(
+            id =        "PUBID_1997",
+            starts_with('CV_BIO_CHILD_HH'),
+            -contains('U18') #Dropping variables with only children under 18
+
+        ) |>
+        pivot_longer(
+            starts_with('CV_BIO_CHILD_HH'),
+            names_to='year',
+            names_pattern='CV_BIO_CHILD_HH_(\\d\\d\\d\\d)',
+            values_to='bio_child_hh',
+            values_transform = list(bio_child_hh = as.integer)
+        )
+
+    data2 = nlsydf |>
+        select(
+            id =        "PUBID_1997",
+            starts_with('CV_BIO_CHILD_NR'),
+            -contains('U18') #Dropping variables with only children under 18
+
+        ) |>
+        pivot_longer(
+            starts_with('CV_BIO_CHILD_NR'),
+            names_to='year',
+            names_pattern='CV_BIO_CHILD_NR_(\\d\\d\\d\\d)',
+            values_to='bio_child_nr',
+            values_transform = list(bio_child_nr = as.integer)
+        )
+
+    data = left_join(data1, data2, by=c('id', 'year')
+    ) |>
+        mutate(bio_child = bio_child_hh + bio_child_nr,
+               year = as.integer(year)
+        )|>
+        group_by(id) |>
+        complete(year=1997:2019) |>
+        fill(bio_child, bio_child_hh, bio_child_nr)|>
+        mutate(total_bio_children = case_when(
```

Comment:
At this point, `bio_child_hh` and `bio_child_nr` should have no missing values and we can just add them to create `bio_child`.

Reply:
Fixed!